[GOMUNGNAM] step-5 게시글 작성, 수정, 삭제 기능 구현 완료

KOKEONHO opened the pull request:
어제 Hyun과 대화하다가 주소창에 사용자의 ID가 노출되는 것에 묘한 위화감을 느낀 것도 있고, 짠 코드를 다시 한번 뜯어보면서 스프링에 익숙해지는 시간을 가지고자, 사용자가 개인 정보를 수정 또는 프로필 열람 할 때 url에 사용자의 ID 대신 사용자의 고유 번호가 노출되도록 수정했습니다. 또한 Honux가 이는 지양하는게 좋을 것 같다고 수업 시간에 말씀하셨던게 떠올라 기존에 작성자를 필드로 지니고 있던 Article 클래스에서도 작성자 필드를 제거했습니다. (자세히 기억은 나지 않지만 그렇게 말씀하셨던 것 같습니다.)

이렇게 바꾸니 잘 와 닿지 않던 5단계의 요구 사항 중 하나였던 '데이터베이스 모델링' 조건도 충족할 수 있게 되었습니다. 원래는 게시글이 작성자 정보를 가지고 있었기 때문에 article 테이블과 member 테이블의 join의 필요성을 느끼지 못했지만, join 없이는 게시글 목록이나 상세 열람 시에 작성자 정보를 가지고 올 수 없음을 깨닫고 두 테이블의 join 공부를 했습니다. 아직 sql문 공부는 더 해야겠다고 느꼈습니다.

짠 코드의 대부분을 수정해야 하는 일이었지만 생각보다 재미있었고 점점 머릿 속에 동작 방식(?)이 그려지는 것 같아 뿌듯했습니다. 불필요한 작업일 수도 있겠지만 막상 하고 나니 얻은게 더 많았습니다!

KOKEONHO committed:
Feat: Constants.java 생성

상수 처리한 변수들을 선언해놓은 Constants.java 생성

KOKEONHO committed:
Feat: LoginController.java 기능 추가

Constants.java에서 상수 선언한 변수를 사용하기 위한 import static 추가

KOKEONHO committed:
Refactor: LoginService.java 코드 수정

기존 반환 값이었던 Member에서 <Optional> Member로 코드 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

Constants.java에서 상수 처리했던 변수로 코드 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

기존 Object로 받아오던 session.getAttribute()를 Member로 형 변환 해준 뒤 변수에 대입하도록 코드 수정

KOKEONHO committed:
Feat: login_failed.html 생성

로그인 실패 시에 보여주는 login_failed.html 생성

KOKEONHO committed:
Feat: PageControllerAdvice.java 생성

Controller에서 발생하는 예외들을 처리해주기 위한 PageControllerAdvice.java 생성

KOKEONHO committed:
Refactor: login_failed.html 코드 수정

login_failed.html 코드 내부의 중복 코드 제거

KOKEONHO committed:
Remove: PageControllerAdvice.java 제거

PageControllerAdvice.java 제거

KOKEONHO committed:
Chore: Constants.java 패키지 이동

기존 Controller 내부에 있던 Constants.java를 상위 패키지로 이동

KOKEONHO committed:
Refactor: Constants.java 패키지 이동으로 인한 코드 수정

Constants.java 패키지 이동으로 인한 import문 코드 수정

KOKEONHO committed:
Feat: LoginController.java 기능 추가

로그인 실패 시 실패했다는 경고 문구가 출력되는 login_failed.html를 보여주는 기능 추가

KOKEONHO committed:
Refactor: login_failed.html 코드 수정

post 요청 시 이동하는 url을 localhost:8080/login으로 지정해주는 코드 수정

KOKEONHO committed:
Feat: ArticleController.java 기능 추가

로그인 하지 않은 상태에서 글을 쓰려고 할 경우 로그인 화면으로 연결해주는 기능 추가

KOKEONHO committed:
Feat: ArticleController.java 기능 추가

로그인 하지 않은 사용자가 게시글을 상세 열람 하려고 하면 로그인 하는 페이지로 이동시키는 기능 추가

KOKEONHO committed:
Refactor: fragment.html 코드 수정

기존 session에 Member 객체를 그대로 넘겨주던 코드에서 Member 객체의 userId만 넘겨주는 코드로 수정함으로써 발생한 코드 수정

KOKEONHO committed:
Refactor: LoginController.java 코드 수정

기존 session에 Member 객체를 그대로 전달해주던 코드에서 Member 객체의 userId만 전달해주는 코드로 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

사용자의 로그인 여부를 토대로 각기 다른 페이지로 이동시켜주는 코드 수정

KOKEONHO committed:
Feat: MemberController.java 기능 추가

로그인하지 않은 사용자의 열람 권한을 축소시키는 기능 추가

KOKEONHO committed:
Refactor: fragment.html 코드 수정

오탈자 수정

KOKEONHO committed:
Feat: show.html 기능 추가

게시글의 작성 시간이 정상적으로 출력되는 기능 추가

KOKEONHO committed:
Feat: form.html 기능 추가

세션에 로그인 된 사용자의 userId를 게시글 작성 시 글쓴이의 값이 들어가는 칸에 자동으로 할당해주는 기능 추가

KOKEONHO committed:
Feat: list.html 기능 추가

사용자가 로그인 되어있는 자신의 계정에만 수정 버튼이 활성화 되는 기능 추가

KOKEONHO committed:
Refactor: list.html 코드 수정

url에 사용자의 ID가 노출되지 않도록 코드 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

url에 사용자의 ID 대신 사용자의 번호가 노출되도록 코드 수정

KOKEONHO committed:
Feat: NamedJdbcTemplateMemberRepository.java 기능 추가

사용자의 번호로 사용자의 정보를 찾는 기능 추가

KOKEONHO committed:
Refactor: schema.sql 코드 수정

table을 drop하는 순서 조정을 위한 코드 수정

KOKEONHO committed:
Refactor: LoginController.java 코드 수정

session이 지니는 값을 userId에서 userSequence로 바꾸는 코드 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

url이 기존에는 userId를 지니고 있었지만 userSequence를 지니게 바꾸면서 발생한 코드 수정

KOKEONHO committed:
Feat: MemberLoginDto.java 기능 추가

필드로 userSequence를 추가하고 이를 생성자를 통해 초기화 가능하게 하는 기능 추가

KOKEONHO committed:
Refactor: updateForm.html 코드 수정

개인 정보 수정을 할 때 수정이 불가능 한 userId 부분에 기존 userId가 정상적으로 출력되게 코드 수정

KOKEONHO committed:
Refactor: list.html 코드 수정

session에서 가지고 있는 userSequence가 일치하는 사용자의 옆에만 수정 버튼이 나타나도록 코드 수정

KOKEONHO committed:
Refactor: schema.sql 코드 수정

member 테이블의 member_id를 unique로 바꾸고 article 테이블에 foreign key로 member 테이블의 member_number를 가리키게 코드 수정

KOKEONHO committed:
Feat: ArticleWithWriterDto.java 생성

게시글의 작성자까지 필드로 담고 리스트에서 보일 때와 상세 열람 시에 정보를 넘겨주는 ArticleWithWriterDto.java 생성

KOKEONHO committed:
Feat: ArticleWithWriterDto.java 생성

게시글의 작성자까지 필드로 담고 리스트에서 보일 때와 상세 열람 시에 정보를 넘겨주는 ArticleWithWriterDto.java 생성

KOKEONHO committed:
Feat: ArticleWithWriterDto.java 기능 추가

기본 생성자 추가

KOKEONHO committed:
Refactor: show.html 코드 수정

userId가 아닌 userSequence가 포함하도록 바뀐 url로 이동이 가능해지게 코드 수정

KOKEONHO committed:
Feat: NamedJdbcTemplateArticleRepository.java 기능 추가

ArticleWithWriterDto를 통해서 필요한 정보를 담은 articleWithWriterDto를 반환할 수 있도록 하는 기능 추가

KOKEONHO committed:
Remove: MemberRepository.java 제거

필요가 없어진 MemberRepository.java 제거

KOKEONHO committed:
Refactor: index.html 코드 수정

userId가 아닌 userSequence가 포함하도록 바뀐 url로 이동이 가능해지게 코드 수정

KOKEONHO committed:
Refactor: form.html 코드 수정

글 작성 화면에서 글쓴이 입력 필드를 삭제하도록 코드 수정

KOKEONHO committed:
Refactor: data.sql 코드 수정

article 테이블에 data.sql을 통해 게시글을 삽입할 때 member_number를 넣어줄 수 없는 문제로 인해 원활한 검증을 위하여 코드 임시 수정

KOKEONHO committed:
Feat: ArticleWithWriterDto.java 기능 추가

ArticleWithWriterDto에 게시글 작성자의 ID를 담을 수 있도록 필드 추가 및 getter, setter 추가

KOKEONHO committed:
Remove: ArticleRepository.java 제거

필요가 없어진 ArticleRepository.java 제거

KOKEONHO committed:
Feat: ArticleController.java 기능 추가

기존 Article 객체를 통해 받아오던 게시글 정보를 새롭게 구현한 ArticleWithWriterDto를 통해 받아와 가공하는 기능 추가

KOKEONHO committed:
Refactor: Article.java 코드 수정

기존에 지니고 있던 writer 필드를 제거하고 userSequence 필드를 추가하는 코드 수정

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

주석 처리했던 코드 제거

KOKEONHO committed:
Refactor: form.html 코드 수정

코드 정리 및 공백 제거 코드 수정

KOKEONHO committed:
Feat: editForm.html 생성

게시글 수정 시에 사용자가 보게되는 editForm.html 생성

KOKEONHO committed:
Feat: ArticleWithWriterDto.java 기능 추가

게시글을 수정할 때 필요한 정보를 담을 ArticleWithWriterDto.java 생성자 추가

KOKEONHO committed:
Refactor: editForm.html 코드 수정

put으로 요청을 보내기 위한 코드 수정

KOKEONHO committed:
Feat: ArticleController.java 기능 추가

게시글을 수정하는 기능 추가

KOKEONHO committed:
Feat: NamedJdbcTemplateArticleRepository.java 기능 추가

게시글을 수정하여 업데이트 하는 기능 추가

KOKEONHO committed:
Feat: ArticleController.java 기능 추가

게시글 삭제 기능 추가

KOKEONHO committed:
Feat: NamedJdbcTemplateArticleRepository.java 기능 추가

게시글 삭제 기능 추가

KOKEONHO committed:
Refactor: show.html 코드 수정

삭제 클릭 시에 해당 url로 보내주도록 코드 수정

KOKEONHO committed:
Refactor: ArticleRepositoryTest.java 코드 수정

임시 주석 처리

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

기존 서브 쿼리를 사용했던 sql문에서 join을 사용하는 sql문으로 코드 수정

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

sql문의 실행을 확인하기 위해 출력을 하던 코드 제거

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

log를 통한 출력으로 확인하기 위한 코드 수정

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

RowMapper 부분의 코드를 람다로 코드 수정

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

쓰이지 않는 import문 제거

honux77 commented:
수고하셨습니다. 공부할 때는 삽질(?)을 통해 원래 얻는 게 많아요 😅

honux77 commented on src/main/java/kr/codesqaud/cafe/Constants.java:
```diff
@@ -0,0 +1,9 @@
+package kr.codesqaud.cafe;
+
+public final class Constants {
+	private Constants() {
+		// restrict instantiation
+	}
```
👍

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/MemberController.java:
```diff
@@ -42,35 +44,45 @@ public String saveUser(@ModelAttribute("user") Member user) throws SQLException
-	@GetMapping("/profile/{userId}")
-	public String userProfile(@PathVariable String userId, Model model) throws SQLException {
-		Optional<Member> findUser = namedJdbcTemplateMemberRepository.findById(userId);
+	@GetMapping("/profile/{userSequence}")
```
근데 아이디나 닉네임이 노출되어도 크게 상관은 없어요. GitHub도 URL에 닉네임을 노출하고 있잖아요? 뭔가 애매할 때 GitHub의 URL을 참고하는 것도 좋은 선택입니다.

KOKEONHO replied:
네 처음에는 주소에 ID가 노출되는게 마냥 이상했다가 Dewey에게서 GitHub도 사용자 ID가 나온다는 말을 듣고 바꾸지 말까 고민도 했습니다. 그런데 일단 학습하는 것이 우선이라는 생각도 들고 제가 짠 코드를 방향을 잡고 고치다보면 스프링이 좀 더 익숙해지지 않을까 싶어서 수정 해봤습니다! 다음에는 Honux 말씀대로 GitHub 주소 참고해봐야겠어요.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/MemberController.java:
```diff
@@ -42,35 +44,45 @@ public String saveUser(@ModelAttribute("user") Member user) throws SQLException
-	public String userProfile(@PathVariable String userId, Model model) throws SQLException {
-		Optional<Member> findUser = namedJdbcTemplateMemberRepository.findById(userId);
+	@GetMapping("/profile/{userSequence}")
+	public String userProfile(HttpSession session, @PathVariable Long userSequence, Model model) throws SQLException {
```
```suggestion
	public String getProfile(HttpSession session, @PathVariable Long userSequence, Model model) throws SQLException {
```

콘트롤러 메소드의 네이밍 컨벤션도 한 번 공부해 보시면 좋을 것 같아요.

KOKEONHO replied:
넵 알겠습니다

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -9,44 +10,95 @@
+
+	@GetMapping("/")
+	public String listArticles(Model model) throws SQLException {
+		List<ArticleWithWriterDto> articles = articleRepository.showAllArticles();
```
```suggestion
		List<ArticleWithWriterDto> articles = articleRepository.findAllArticlesWithWriter();
```

저도 좋은 이름은 잘 모르겠지만 레포의 네이밍 컨벤션도 학습해 보시면 좋을 것 같아요.

KOKEONHO replied:
네 안 그래도 이름 붙이는게 생각보다 어렵더군요...

honux77 commented on src/main/java/kr/codesqaud/cafe/dto/ArticleWithWriterDto.java:
```diff
@@ -0,0 +1,80 @@
+
+	private Long articleSequence;
+	private String title;
+	private String contents;
```
GET questions/list 에서 이 DTO를 이용하는데 contents가 없는 편이 더 효율적이지 않을까요?

KOKEONHO replied:
게시글 목록을 출력하는 questions/list에서도 이 DTO가 쓰이지만 게시글 상세 열람을 할 때에도 이 DTO가 쓰여서 contents를 뺄 수가 없습니다. 혹시 여러개의 DTO를 사용해도 상관이 없는건가요 ? 예를 들어서 questions/list에는 굳이 contents가 필요 없으니, contents가 없는 ArticleWithoutContentsDto를 만들어서 사용한다든지 ...

honux77 replied:
맞아요. 여러개를 사용해도 상관없습니다. 뺄 수 없으면 따로 만들어야죠.

honux77 commented on src/main/java/kr/codesqaud/cafe/repository/NamedJdbcTemplateArticleRepository.java:
```diff
@@ -22,35 +20,57 @@ public NamedJdbcTemplateArticleRepository(NamedParameterJdbcTemplate template) {
-		String sql = "select article_number, article_writer, article_title, article_contents, article_writtentime from article";
-		return template.query(sql, articleRowMapper());
+	public List<ArticleWithWriterDto> showAllArticles() {
+		String sql = "select a.article_number, a.article_title, a.article_contents, a.article_writtentime, a.member_number, m.member_id as writer, " +
```
여기도 마찬가지 contents가 없으면 좋을 것 같아요.

KOKEONHO replied:
DTO를 새롭게 만들어서 해봤습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -9,44 +10,95 @@
+		return "index";
+	}
+
+	@GetMapping("/qna/questions")
```
```suggestion
	@GetMapping("/qna/questions/form")
```

KOKEONHO replied:
수정했습니다

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -9,44 +10,95 @@
+
+	@PostMapping("/qna/questions")
+	public String saveQuestion(@RequestParam String title, @RequestParam String contents, HttpSession session) throws SQLException {
+		Article article = new Article((Long) session.getAttribute(SESSIONED_USER), title, contents);
```
session.getAttribute(SESSIONED_USER) 코드가 계속 중복됩니다. 요것도 줄일 수 있을까요?

KOKEONHO replied:
메서드로 추출해서 코드 수정해봤습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -9,44 +10,95 @@
+	}
+
+	@PostMapping("/qna/questions")
+	public String saveQuestion(@RequestParam String title, @RequestParam String contents, HttpSession session) throws SQLException {
```
여기서도 로그인 되었는지 여부를 확인해야 하는데 안 하고 있네요.

KOKEONHO replied:
LoginInterceptor 생성해서 로그인 여부 확인하게 해봤습니다.

honux77 replied:
앗 그렇군요.

honux77 reviewed (changes requested):
수고하셨습니다.

1) URL 규칙을 학습하고 수정해 볼 것 
2) 콘트롤러 및 레포  메소드 이름 개선 
3) 필요한 모든 곳에 로그인 여부 검증, API 권한 확인 코드가 잘 되었는지 확인할 것

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

Honux의 피드백을 바탕으로 기존 showAllArticles()였던 메서드 명을 findAllArticlesWithWriter()로 수정

KOKEONHO committed:
Refactor: ArticleRepositoryTest.java 코드 수정

Honux의 피드백을 바탕으로 기존 showAllArticles()였던 메서드 명을 findAllArticlesWithWriter()로 수정

KOKEONHO committed:
Feat: ArticleWithoutContentsDto.java 생성

게시글 목록에서는 게시글의 contents가 불필요하기 때문에 contents를 가지고 있지 않은 게시글 DTO인 ArticleWithoutContentsDto.java 생성

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

Honux의 피드백을 바탕으로 기존 userProfile()였던 메서드 명을 gerProfile()로 수정

KOKEONHO committed:
Refactor: NamedJdbcTemplateArticleRepository.java 코드 수정

Honux의 피드백을 바탕으로 기존 showAllArticles()였던 메서드 명을 findAllArticlesWithWriter()로 수정

KOKEONHO committed:
Refactor: ArticleWithWriterDto.java 코드 수정

쓰이지 않은 불필요한 생성자 제거

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

새롭게 생성한 ArticleWithoutContentsDto로 인한 메서드 명 변경으로 코드 수정

KOKEONHO committed:
Feat: ArticleWithoutContentsDto.java 생성

게시글 목록에서는 게시글의 contents가 불필요하기 때문에 contents를 가지고 있지 않은 게시글 DTO인 ArticleWithoutContentsDto.java 생성

KOKEONHO committed:
Feat: NamedJdbcTemplateArticleRepository.java 기능 추가

contents가 빠진 게시글 DTO를 사용하기 위한 sql문 수정 및 메서드 명 변경

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

게시글을 쓰는 URL을 기존 /qna/questions에서 /qna/questions/form으로 코드 수정

KOKEONHO committed:
Refactor: index.html 코드 수정

바뀐 URL을 반영한 코드 수정

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

LoginInterceptor.java 생성에 따른 기존 세션 확인 코드 수정

KOKEONHO committed:
Feat: LoginInterceptor.java 생성

로그인 권한 확인이 필요할 때에 컨트롤러 호출 이전 이를 확인해주는 LoginInterceptor.java 생성

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

LoginInterceptor.java 생성에 따른 기존 세션 확인 코드 수정

KOKEONHO committed:
Feat: WebMvcConfig.java 생성

새롭게 추가한 LoginInterceptor.java를 설정해주는 WebMvcConfig.java 생성

KOKEONHO committed:
Refactor: ArticleController.java 코드 수정

중복되는 session.getAttribute(SESSIONED_USER) 코드를 메서드로 추출하여 코드 수정

KOKEONHO committed:
Refactor: MemberController.java 코드 수정

중복되는 session.getAttribute(SESSIONED_USER) 코드를 메서드로 추출하여 코드 수정
